lib/AutocorrectingInterval.ts

```diff
@@ -0,0 +1,53 @@
+export type AutocorrectingInterval = number
+export type TimeoutFunction = (
+  callback: () => void,
+  millis: number
+) => NodeJS.Timeout
+
+let currentId = 0 as AutocorrectingInterval
+const intervals = new Map<number, NodeJS.Timeout>()
+
+/**
+ * Sets an interval that automatically adjusts the delay between each
+ * interval tick by the average of all sleep inaccuracies.
+ * (eg. If one interval tick sleeps for 50ms more than the given `millis`,
+ * then the next duration tick will be adjusted to
+ * `millis - (previousAverageOffset + 50) / 2`.)
+ *
+ * @returns An identifier that should be passed to
+ * {@link clearAutocorrectingInterval} that cancels the interval.
+ */
+export const setAutocorrectingInterval = (
+  callback: () => void,
+  millis: number,
+  timeoutFn: TimeoutFunction = setTimeout
+): AutocorrectingInterval => {
+  const id = currentId++
+  let nowTime = Date.now()
+  let currentIntervalMillis = millis
+  let averageSleepOffset: number | undefined
+  const run = () => {
+    callback()
+    const endTime = Date.now()
+    const currentSleepOffset = endTime - nowTime - currentIntervalMillis
+    averageSleepOffset = averageSleepOffset
+      ? (averageSleepOffset + currentSleepOffset) / 2
+      : currentSleepOffset
+    nowTime = endTime
```

Comment:
Instead of updating the nowTime in each iteration, we could use the initial nowTime as an anchor point to eliminate the possibility of compounding error. So instead of adjusting each interval to last as close to one second as possible, we would adjust each interval to align our new time as close as possible to what should be the "real time". Something like this:

```suggestion
export const setAutocorrectingInterval = (
  callback: () => void,
  millis: number,
  timeoutFn: TimeoutFunction = setTimeout
): AutocorrectingInterval => {
  const id = currentId++;
  const startTime = Date.now();
  let intervalCount = 0;
  const run = () => {
    callback();
    intervalCount++;
    const expectedTime = startTime + intervalCount * millis;
    const drift = Date.now() - expectedTime;
    const nextIntervalMillis = millis - drift;
    intervals.set(id, timeoutFn(run, nextIntervalMillis));
  };
  intervals.set(id, timeoutFn(run, millis));
  return id;
};
```

Reply:
I gave this a try, but it didn't seem to have many benefits over the average-based scheme. In fact, I think it responds to outliers slightly worse, but overall there's not much of a difference. Here's some data from an iPhone 15 Pro (I had similar results on other devices):
```
-- True Average (iPhone 15 Pro)

Offset 26 Sleep Duration 1026 Current Interval 1000
Offset 28 Sleep Duration 1002 Current Interval 974
Offset 26 Sleep Duration 999 Current Interval 973
Offset 27.66666666666663 Sleep Duration 1001 Current Interval 973.3333333333334
Offset 24.91666666666663 Sleep Duration 998 Current Interval 973.0833333333334
Offset 31.51666666666665 Sleep Duration 1005 Current Interval 973.4833333333333
Offset 22.350000000000023 Sleep Duration 995 Current Interval 972.65
Offset 27.635714285714243 Sleep Duration 1001 Current Interval 973.3642857142858
Offset 25.760714285714243 Sleep Duration 999 Current Interval 973.2392857142858
Offset 28.649603174603158 Sleep Duration 1002 Current Interval 973.3503968253968
Offset 25.849603174603203 Sleep Duration 999 Current Interval 973.1503968253968
Offset 25.75869408369408 Sleep Duration 999 Current Interval 973.2413059163059
Offset 28.67536075036071 Sleep Duration 1002 Current Interval 973.3246392496393
Offset 29.8292069042069 Sleep Duration 1003 Current Interval 973.1707930957931
Offset 24.043492618492564 Sleep Duration 997 Current Interval 972.9565073815074
Offset 26.843492618492633 Sleep Duration 1000 Current Interval 973.1565073815074
Offset 27.843492618492633 Sleep Duration 1001 Current Interval 973.1565073815074
Offset 27.902316147904344 Sleep Duration 1001 Current Interval 973.0976838520957
Offset 24.957871703459887 Sleep Duration 998 Current Interval 973.0421282965401
Offset 26.852608545565204 Sleep Duration 1000 Current Interval 973.1473914544348
Offset 30.852608545565204 Sleep Duration 1004 Current Interval 973.1473914544348
Offset 21.04308473604135 Sleep Duration 994 Current Interval 972.9569152639586
Offset 13.770357463314099 Sleep Duration 987 Current Interval 973.2296425366859
Offset 27.205140072009726 Sleep Duration 1001 Current Interval 973.7948599279903
Offset 24.24680673867647 Sleep Duration 998 Current Interval 973.7531932613235
Offset 24.166806738676428 Sleep Duration 998 Current Interval 973.8331932613236
Offset 29.089883661753333 Sleep Duration 1003 Current Interval 973.9101163382467
Offset 26.20099477286442 Sleep Duration 1000 Current Interval 973.7990052271356
Offset 27.20099477286442 Sleep Duration 1001 Current Interval 973.7990052271356
Offset 33.23547753148512 Sleep Duration 1007 Current Interval 973.7645224685149
Offset 21.468810864818465 Sleep Duration 995 Current Interval 973.5311891351815
Offset 25.307520542237853 Sleep Duration 999 Current Interval 973.6924794577621
Offset 25.276270542237853 Sleep Duration 999 Current Interval 973.7237294577621
Offset 28.24596751193485 Sleep Duration 1002 Current Interval 973.7540324880651
Offset 27.30479104134656 Sleep Duration 1001 Current Interval 973.6952089586534
Offset 24.33336246991803 Sleep Duration 998 Current Interval 973.666637530082
Offset 29.277806914362486 Sleep Duration 1003 Current Interval 973.7221930856375
Offset 41.358887995443524 Sleep Duration 1015 Current Interval 973.6411120045565
Offset 26.75362483754884 Sleep Duration 1000 Current Interval 973.2463751624512
Offset 26.75362483754884 Sleep Duration 1000 Current Interval 973.2463751624512

-- Suggestion (iPhone 15 Pro)

Offset 27 Sleep Duration 1027 Current Interval 1000
Offset 28 Sleep Duration 1001 Current Interval 973
Offset 26 Sleep Duration 998 Current Interval 972
Offset 28 Sleep Duration 1002 Current Interval 974
Offset 27 Sleep Duration 999 Current Interval 972
Offset 33 Sleep Duration 1006 Current Interval 973
Offset 29 Sleep Duration 996 Current Interval 967
Offset 27 Sleep Duration 998 Current Interval 971
Offset 28 Sleep Duration 1001 Current Interval 973
Offset 18 Sleep Duration 990 Current Interval 972
Offset 29 Sleep Duration 1011 Current Interval 982
Offset 28 Sleep Duration 999 Current Interval 971
Offset 30 Sleep Duration 1002 Current Interval 972
Offset 28 Sleep Duration 998 Current Interval 970
Offset 34 Sleep Duration 1006 Current Interval 972
Offset 31 Sleep Duration 997 Current Interval 966
Offset 30 Sleep Duration 999 Current Interval 969
Offset 30 Sleep Duration 1000 Current Interval 970
Offset 30 Sleep Duration 1000 Current Interval 970
Offset 31 Sleep Duration 1001 Current Interval 970
Offset 30 Sleep Duration 999 Current Interval 969
Offset 15 Sleep Duration 985 Current Interval 970
Offset 30 Sleep Duration 1015 Current Interval 985
Offset 20 Sleep Duration 990 Current Interval 970
Offset 15 Sleep Duration 995 Current Interval 980
Offset 15 Sleep Duration 1000 Current Interval 985
Offset 15 Sleep Duration 1000 Current Interval 985
```
The big question in doing this is how do we predict the next offset such that we can set the next sleep duration to ensure the actual sleep duration is 1000ms. More or less, we need to know the tendencies of the device clock above all else.

Overall I think the most accurate solution would be to dip into native code, but I don't want to do that for a simple countdown timer.